Reduce CurlHandler header parse allocations

justinvp opened the pull request:
Use existing known header and reason phrase instances instead of always allocating new strings (similar to #3199 for WinHttpHandler) and avoid unnecessary intermediate string allocations.

cc: @stephentoub, @kapilash, @davidsh

(@stephentoub, is there a good way to profile the allocations for a before/after comparison on Unix?)

Note: I only tested this on OS X as I didn't have a Linux environment readily available; relying on CI for the other *nixes.

stephentoub commented:
Thanks, @justinvp!

> @stephentoub, is there a good way to profile the allocations for a before/after comparison on Unix?

PerfView has been updated with support for data captured on Linux:
http://blogs.msdn.com/b/vancem/archive/2016/02/20/analyzing-cpu-traces-from-linux-with-perfview.aspx
but I don't believe it's yet at a point where it can help with allocation analysis, though at some point it will be able to.  In the meantime, you can do something very rudimentary and macro by looking at GC.CollectionCount(0) for example before and after some perf test.

stephentoub commented on src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs:
```diff
@@ -226,5 +271,31 @@ private static bool TryMatch(string known, char[] key, int startIndex, int lengt
+            return true;
+        }
+
+        private delegate char CharAt<T>(T key, int index);
```
Why not ```Func<T, int, char>```?  Similar question for the delegate below?

justinvp replied:
I originally had `Func<...>` but changed them to concrete delegates to make the parameters more clear. Though, it doesn't matter all that much here. I can change it back to `Func<...>`s if you want.

stephentoub replied:
I always have a small preference for not introducing new delegate types, but if you think it leads to the code being more understandable, it's ok to leave as-is.

justinvp replied:
Switched back to `Func`.

stephentoub commented on src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs:
```diff
@@ -12,191 +12,234 @@ internal static partial class HttpKnownHeaderNames
+        {
+            Debug.Assert(length >= 0);
+
+            if (buffer == IntPtr.Zero)
```
Is length == 0 common?  I assume not, but if it is you could add the check here as well.

justinvp replied:
I don't think `length == 0` is common. Honestly, `buffer == IntPtr.Zero` isn't common either in practice -- I originally asserted that `buffer != IntPtr.Zero` but found in testing that this assert would fire when I would test that `string.Empty` wasn't found, due to `fixed` on an empty array returning a null pointer. I could have special-cased the `string.Empty` test to avoid this, but figured I might as well handle `buffer == IntPtr.Zero` in `TryGetHeaderName` in case it is ever called with a null pointer. The `length == 0` case is already handled, just not early here, but in the subsequent call.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:
```diff
@@ -603,17 +605,28 @@ private static ulong CurlReceiveHeadersCallback(IntPtr buffer, ulong size, ulong
                         HttpResponseMessage response = easy._responseMessage;
 
-                        if (!TryParseStatusLine(response, responseHeader, easy))
+                        CurlResponseHeaderReader reader = new CurlResponseHeaderReader(buffer, size);
```
Are headers guaranteed to be ASCII?  It looks like your indexing under that assumption, but the previous code was using Marshal.PtrToStringAnsi, which is UTF8 on Unix.

davidsh replied:
Please review RFC 7230:
http://tools.ietf.org/html/rfc7230#page-22

davidsh replied:
Also this RFC:
http://tools.ietf.org/html/rfc5322

As you can here:

>2.2.  Header Fields
   Header fields are lines beginning with a field name, followed by a
   colon (":"), followed by a field body, and terminated by CRLF.  A
   field name MUST be composed of printable US-ASCII characters (i.e.,
   characters that have values between 33 and 126, inclusive), except
   colon.  A field body may be composed of printable US-ASCII characters
   as well as the space (SP, ASCII value 32) and horizontal tab (HTAB,
   ASCII value 9) characters (together known as the white space
   characters, WSP).  A field body MUST NOT include CR and LF except
   when used in "folding" and "unfolding", as described in section
   2.2.3.  All field bodies MUST conform to the syntax described in
   sections 3 and 4 of this specification.

stephentoub replied:
Thanks, David.  So treating it as ASCII is fine.

justinvp replied:
Thanks for the pointers to the RFCs, David. I referenced RFC 7230 while working on this. The status line and headers chars we're parsing should be ASCII per the RFC. According to RFC 7230, it is possible that a header value has chars outside of ASCII:

> 3.2.4.  Field Parsing
> Historically, HTTP has allowed field content with text in the
   ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
   through use of [RFC2047] encoding.  In practice, most HTTP header
   field values use only a subset of the US-ASCII charset [USASCII].
   Newly defined header fields SHOULD limit their field values to
   US-ASCII octets.  A recipient SHOULD treat other octets in field
   content (obs-text) as opaque data.

 ...but we're not parsing the header values (other than a `Trim` on the edges) and just passing off the header value substring to `Marshal.PtrToStringAnsi` as was being done before this PR.

justinvp replied:
Which makes me wonder: should [`HttpRuleParser.DefaultHttpEncoding`](https://github.com/dotnet/corefx/blob/11efc0d43d56e1c280653934b27ae5c1ef05af8c/src/System.Net.Http/src/System/Net/Http/HttpRuleParser.cs#L42-L47) be used to materialize the header value instead of `Marshal.PtrToStringAnsi`?

stephentoub replied:
Probably.

justinvp replied:
Let me do some more testing to see what curl hands back for non-ASCII header values. Probably won't get to it until later today.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseHeaderReader.cs:
```diff
@@ -0,0 +1,282 @@
+        public CurlResponseHeaderReader(IntPtr buffer, ulong size)
+        {
+            Debug.Assert(buffer != IntPtr.Zero);
+            // In actuality, size will be <= CURL_MAX_HTTP_HEADER (1024*100) bytes.
```
Then should we assert that?

justinvp replied:
I added an assert for `CURL_MAX_HTTP_HEADER` in the the callback method. I was going to add it here, but decided it made more sense to define `CURL_MAX_HTTP_HEADER` with the other interop definitions and didn't want `CurlResponseHeaderReader` to depend on the interop because `CurlResponseHeaderReader` is tested in the `System.Net.Http` unit tests that don't currently depend on any interop file and I didn't want to have to pull the interop files into the unit tests just for this.

I left the `int.MaxValue` assert here because that's really all we care about here.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseHeaderReader.cs:
```diff
@@ -0,0 +1,282 @@
+                get
+                {
+                    Debug.Assert(index >= 0 && index < Length);
+                    return _pointer[index];
```
Maybe also assert that ```_pointer[index] <= 0x7F```?

justinvp replied:
Done.

stephentoub commented:
This is very nice, Justin.  Thanks for doing this.  A few nits, otherwise LGTM.

justinvp commented:
Thanks, Steve and David. I added another PR that addresses the feedback (and a few other minor nits I noticed when going over the code again). I will squash before merging.

I ran the following before/after on OS X:

```csharp
using (var client = new HttpClient())
{
    for (int i = 0; i < 1000; i++)
    {
        client.GetAsync("http://httpbin.org/ip").GetAwaiter().GetResult().Dispose();
    }
    Console.WriteLine(GC.CollectionCount(0));
}
```

Before: 4
After: 3

I suppose if I increased the iterations we might see a more dramatic drop in Gen0 GCs.

kapilash commented:
This is great, Thanks @justinvp

stephentoub commented:
Test Innerloop CentOS7.1 Debug Build and Test please
Test Innerloop OSX Debug Build and Test please

justinvp commented:
We definitely want to be using Latin1 encoding (`HttpRuleParser.DefaultHttpEncoding`) over `Marshal.PtrToStringAnsi` to properly carry forward non-US-ASCII octets, to be consistent with `WinHttpHandler` and the RFC. I've updated the "Address PR feedback" commit.

I have found inconsistencies between libcurl and WinHTTP with how WinHTTP handles invalid chars. libcurl doesn't do much validation at all. It mostly just passes each header line to the header callback and in our callback we don't add much validation ourselves. For example, WinHTTP fails if a null char is in a header value, but libcurl doesn't complain (nor do we). I'll open a separate PR soon (with tests) to address this.

Let me know if the changes in this PR look good and I'll squash into a single commit.

davidsh commented:
>I have found inconsistencies between libcurl and WinHTTP with how WinHTTP handles invalid chars. libcurl doesn't do much validation at all. It mostly just passes each header line to the header callback and in our callback we don't add much validation ourselves. For example, WinHTTP fails if a null char is in a header value, but libcurl doesn't complain (nor do we). I'll open a separate PR soon (with tests) to address this.

If you have time, it would be a good datapoint to get similar test results on validation from .NET Framework (Desktop) using HttpClient (HttpClientHandler).  That uses the original managed sockets stack.

stephentoub commented:
> I've updated the "Address PR feedback" commit.

Thanks!

> I'll open a separate PR soon (with tests) to address this.

Sounds good. Thanks.

stephentoub commented:
> Let me know if the changes in this PR look good and I'll squash into a single commit.

LGTM.  No need to squash; the two commits is fine.
